File: app14_LUA_UNITY/backend/db/migrations/003_create_cards.lua
local schema = require("lapis.db.schema")
local types = schema.types
local db = require("lapis.db")

-- D-03/§4: card_kind has no compiler-enforced sum type in Lua (no ADTs, no
-- interfaces at all) — the CHECK constraint below is the ONLY guarantee that
-- a design_harm row never carries a severity. See PLAN.md D-03.
return {
  up = function()
    schema.create_table("cards", {
      { "card_id", types.varchar({ primary_key = true }) },
      { "suit_code", types.varchar({ null = false }) },
      { "suit_name", types.varchar({ null = false }) },
      { "edition", types.varchar({ null = false }) },
      { "value", types.varchar({ null = false }) },
      { "card_kind", types.varchar({ null = false, default = "technical_threat" }) },
      { "severity", types.varchar({ null = true }) },
      { "description_en", types.text({ null = false }) },
      { "description_pl", types.text({ null = false, default = "" }) },
      { "misc_note", types.text({ null = false, default = "" }) },
      { "source_url", types.varchar({ null = false, default = "" }) },
      { "owasp_refs", "text[] not null default '{}'" },
      { "mitre_refs", "text[] not null default '{}'" },
      { "content_sha256", types.varchar({ null = false }) },
      { "is_critical", types.boolean({ null = false, default = false }) }
    })

    db.query("alter table cards add constraint cards_kind_check " ..
      "check (card_kind in ('technical_threat', 'design_harm'))")
    db.query("alter table cards add constraint cards_design_harm_severity_check " ..
      "check ((card_kind = 'design_harm' and severity is null) or " ..
      "(card_kind = 'technical_threat' and severity is not null))")
    db.query("create index cards_edition_idx on cards (edition)")
    db.query("create index cards_suit_code_idx on cards (suit_code)")
  end,
  down = function()
    schema.drop_table("cards")
  end
}

File: app13_ruby_FastApi/backend/db/migrations/003_create_cards.rb
# frozen_string_literal: true

Sequel.migration do
  change do
    create_table(:cards) do
      String :card_id, primary_key: true
      String :suit_code, null: false
      String :suit_name, null: false
      String :edition, null: false
      String :value, null: false
      String :card_kind, null: false
      String :severity # NULL-able — see D-03 constraint below
      String :description_en, text: true, null: false
      String :description_pl, text: true, null: false, default: ""
      String :misc_note, text: true
      String :source_url
      column :owasp_refs, "text[]", null: false, default: Sequel.lit("'{}'")
      column :mitre_refs, "text[]", null: false, default: Sequel.lit("'{}'")
      String :content_sha256, null: false
      TrueClass :is_critical, null: false, default: false

      index :suit_code
      index :edition

      # D-03: it is structurally impossible, at the database level, for a
      # design_harm card to carry a severity, or for a technical_threat card
      # to lack one — the runtime-enforced analogue of the compiler-enforced
      # sum types every statically-typed sibling in this series gets instead.
      constraint(:cards_kind_check) do
        Sequel.lit("card_kind IN ('technical_threat', 'design_harm')")
      end
      constraint(:cards_severity_matches_kind_check) do
        Sequel.lit(
          "(card_kind = 'design_harm' AND severity IS NULL) OR " \
          "(card_kind = 'technical_threat' AND severity IS NOT NULL)"
        )
      end
    end
  end
end
